### ISSUE TITLE
`ImportError: No module named 'encodings'` when using with virtualenv + python3

### ISSUE THREAD

**Issue by yuvipanda** [JuliaPy/PyCall.jl#410]
Hello!

Am trying to install and use this with a Python3 that's in a virtualenv. Am getting the following error:

```
INFO: Initializing package repository /root/.julia/v0.6
INFO: Cloning METADATA from https://github.com/JuliaLang/METADATA.jl
INFO: Cloning cache of BinDeps from https://github.com/JuliaLang/BinDeps.jl.git
INFO: Cloning cache of Compat from https://github.com/JuliaLang/Compat.jl.git
INFO: Cloning cache of Conda from https://github.com/JuliaPy/Conda.jl.git
INFO: Cloning cache of JSON from https://github.com/JuliaIO/JSON.jl.git
INFO: Cloning cache of MacroTools from https://github.com/MikeInnes/MacroTools.jl.git
INFO: Cloning cache of PyCall from https://github.com/JuliaPy/PyCall.jl.git
INFO: Cloning cache of SHA from https://github.com/staticfloat/SHA.jl.git
INFO: Cloning cache of URIParser from https://github.com/JuliaWeb/URIParser.jl.git
INFO: Installing BinDeps v0.6.0
INFO: Installing Compat v0.26.0
INFO: Installing Conda v0.5.3
INFO: Installing JSON v0.12.0
INFO: Installing MacroTools v0.3.7
INFO: Installing PyCall v1.14.0
INFO: Installing SHA v0.3.3
INFO: Installing URIParser v0.1.8
INFO: Building Conda
INFO: Building PyCall
INFO: PyCall is using python (Python 3.5.3) at /srv/venv/bin/python, libpython = libpython3.5m
INFO: /root/.julia/v0.6/PyCall/deps/deps.jl has been updated
INFO: /root/.julia/v0.6/PyCall/deps/PYTHON has been updated
INFO: Package database updated
Fatal Python error: Py_Initialize: Unable to get the locale encoding
ImportError: No module named 'encodings'

Current thread 0x00007f6916239c40 (most recent call first):

signal (6): Aborted
while loading no file, in expression starting on line 0
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
Py_FatalError at /usr/lib/x86_64-linux-gnu/libpython3.5m.so (unknown line)
unknown function (ip: 0x7f68f29c7419)
_Py_InitializeEx_Private at /usr/lib/x86_64-linux-gnu/libpython3.5m.so (unknown line)
__init__ at /root/.julia/v0.6/PyCall/src/pyinit.jl:42
unknown function (ip: 0x7f68f792511f)
jl_call_fptr_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:339 [inlined]
jl_call_method_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:358 [inlined]
jl_apply_generic at /home/centos/buildbot/slave/package_tarball64/build/src/gf.c:1933
jl_apply at /home/centos/buildbot/slave/package_tarball64/build/src/julia.h:1424 [inlined]
jl_module_run_initializer at /home/centos/buildbot/slave/package_tarball64/build/src/toplevel.c:87
jl_init_restored_modules at /home/centos/buildbot/slave/package_tarball64/build/src/dump.c:2443 [inlined]
_jl_restore_incremental at /home/centos/buildbot/slave/package_tarball64/build/src/dump.c:3272
jl_restore_incremental at /home/centos/buildbot/slave/package_tarball64/build/src/dump.c:3292
_include_from_serialized at ./loading.jl:157
_require_from_serialized at ./loading.jl:200
_require at ./loading.jl:491
require at ./loading.jl:398
unknown function (ip: 0x7f690ffbc92b)
jl_call_fptr_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:339 [inlined]
jl_call_method_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:358 [inlined]
jl_apply_generic at /home/centos/buildbot/slave/package_tarball64/build/src/gf.c:1933
jl_apply at /home/centos/buildbot/slave/package_tarball64/build/src/julia.h:1424 [inlined]
eval_import_path_ at /home/centos/buildbot/slave/package_tarball64/build/src/toplevel.c:403
eval_import_path at /home/centos/buildbot/slave/package_tarball64/build/src/toplevel.c:430 [inlined]
jl_toplevel_eval_flex at /home/centos/buildbot/slave/package_tarball64/build/src/toplevel.c:495
jl_toplevel_eval_flex at /home/centos/buildbot/slave/package_tarball64/build/src/toplevel.c:551
jl_toplevel_eval_in at /home/centos/buildbot/slave/package_tarball64/build/src/builtins.c:496
eval at ./boot.jl:235
unknown function (ip: 0x7f690fe9039f)
jl_call_fptr_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:339 [inlined]
jl_call_method_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:358 [inlined]
jl_apply_generic at /home/centos/buildbot/slave/package_tarball64/build/src/gf.c:1933
process_options at ./client.jl:286
_start at ./client.jl:371
unknown function (ip: 0x7f690fed26e8)
jl_call_fptr_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:339 [inlined]
jl_call_method_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:358 [inlined]
jl_apply_generic at /home/centos/buildbot/slave/package_tarball64/build/src/gf.c:1933
unknown function (ip: 0x401e41)
unknown function (ip: 0x401662)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x40170c)
Allocations: 11115444 (Pool: 11107376; Big: 8068); GC: 27
Aborted (core dumped)
```

You can reproduce this with the following Dockerfile:

```
FROM ubuntu:17.04

RUN apt-get update 

RUN apt-get install --yes python3 python3-dev curl python3-venv
ENV JULIA_VERSION 0.6.0
ENV JULIA_PATH /srv/julia
ENV PATH ${JULIA_PATH}/bin:/srv/venv/bin/:${PATH}

RUN mkdir ${JULIA_PATH}
RUN  curl -sSL "https://julialang-s3.julialang.org/bin/linux/x64/${JULIA_VERSION%[.-]*}/julia-${JULIA_VERSION}-linux-x86_64.tar.gz" | tar -xz -C ${JULIA_PATH} --strip-components 1

RUN python3 -m venv /srv/venv

RUN /bin/bash -c "source /srv/venv/bin/activate && julia  -e 'Pkg.add(\"PyCall\"); using PyCall;'"

```

It worked under the following circumstances:

1. When we were just using system python, without a virtualenv
2. When we were using a virtualenv, but with python2
3. When using conda

It did not work under the following variants either:

1. When using fedora instead of ubuntu base image
2. When using python3.6 instead of python3.5 (with Ubuntu)

**Comment by yuvipanda** [comment]
(Thanks to @tkelman for helping debug this to this point!)

**Comment by stevengj** [comment]
See https://stackoverflow.com/questions/38132755/importerror-no-module-named-encodings

**Comment by stevengj** [comment]
See also dagwieers/unoconv#255, which suggests that it may be a problem with your PYTHONPATH environment variable.

**Comment by yuvipanda** [comment]
```
const python = "python"
const libpython = "libpython3.5m"
const pyprogramname = "/srv/venv/bin/python"
const wpyprogramname = Base.cconvert(Cwstring, "/srv/venv/bin/python")
const pyversion_build = v"3.5.3"
const PYTHONHOME = "/srv/venv:/srv/venv"
const wPYTHONHOME = Base.cconvert(Cwstring, "/srv/venv:/srv/venv")

"True if we are using the Python distribution in the Conda package."
const conda = false
```

I am thinking PYTHONHOME is set wrong.

**Comment by yuvipanda** [comment]
I can reproduce this by just running

```
 python3 -c "import distutils.sysconfig; print(distutils.sysconfig.get_config_var('VERSION'))"
```

with the PYTHONHOME that's set by PyCall. If PYTHONHOME is not set at all then it works just fine - python looks at `sys.prefix` to figure out where to look for libraries.

**Comment by yuvipanda** [comment]
If we comment out all setting of PYTHONHOME, it works fine (see https://github.com/yuvipanda/PyCall.jl/tree/homeless branch).

**Comment by yuvipanda** [comment]
Seems to be result of #299

**Comment by AnjaneyuluBatta505** [comment]
https://stackoverflow.com/questions/38132755/importerror-no-module-named-encodings

**Comment by joachimneu** [comment]
Thank you @yuvipanda and @tkelman for debugging this and suggesting a solution. Until this is properly fixed in PyCall, here is a very hacky quick fix to "just make it work":
1. Install PyCall (many users might get it through PyPlot)
2. Patch `v0.6/PyCall/deps/deps.jl` (you might find it in `~/.julia`, or a project-specific `.playground`): comment out `const PYTHONHOME = "..."`, and add `const PYTHONHOME = ""` instead
3. Now `using PyCall` should work

**Comment by stevengj** [comment]
> python looks at sys.prefix to figure out where to look for libraries.

PYTHONHOME *is* set based on `sys.prefix` in [build.jl](https://github.com/JuliaPy/PyCall.jl/blob/master/deps/build.jl#L217) unless you have the PYTHONHOME environment variable set when you are building PyCall.

Can you try `ENV["PYTHONHOME"]=""; Pkg.build("PyCall")`?

**Comment by joachimneu** [comment]
I tried this, it does not work: It shows the error about missing "encodings" module during build, concludes that "No system-wide Python was found", and proceeds to install miniconda. Which I abort, because I have a perfectly working Python as part of the virtualenv, that I want it to use.

**Comment by stevengj** [comment]
What is you `PYTHONHOME` environment variable during the build?

**Comment by stevengj** [comment]
Actually, probably you need to do `pop!(ENV, "PYTHONHOME")` to actually delete the environment variable, rather than setting it to the empty string.

**Comment by joachimneu** [comment]
The python virtualenv does not set `PYTHONHOME`, there is no such key in `ENV` when I start `julia` within the python virtualenv.

**Comment by stevengj** [comment]
Then maybe you can add some `println` statements to the build script to see why `sys.prefix` is not giving the correct PYTHONHOME?

**Comment by aterenin** [comment]
Same issue. Can be replicated by installing the FireDrake virtualenv (link: https://www.firedrakeproject.org/download.html) and trying to get Julia to use it.

**Comment by aterenin** [comment]
Confirmed that the following fixes the issue.

1. Activate virtualenv
2. Set `ENV["PYTHON"]` to correct Python path under the virtualenv
3. Install PyCall
4. Open `v0.6/PyCall/deps/deps.jl` with a text editor and replace `const PYTHONHOME = "..."` with `const PYTHONHOME = ""`
5. Run `using PyCall`

**Comment by PABourdin** [comment]
I just notice that the issue is still present in latest Julia 1.2.0.

When I install "PyCall" with my Python3 installation not in the default place, I have to manually set:
ENV["PYTHON"]="/my/path/to/python-install/bin/python3"
ENV["PYTHONHOME"]="/my/path/to/python-install/bin"
inside a Julia session, so that I may execute:
Pkg.build("PyCall")
Pkg.build("PyPlot")

Is it possible to fix this issue somehow in the Julia/PyCall code, so that installation becomes doable for non-experts? Thank you.

**Comment by PABourdin** [comment]
PS: Can we re-open this issue or shall I file a new one?

**Comment by jriesen** [comment]
With a fresh install of Python 3.10.8 through pyenv and fresh Julia 1.8.2, I also encountered this error after trying to install PyCall from the package manager.  I solved the problem by editing `deps.jl` as described above by @aterenin and setting `PYTHONHOME=""`.  Note that I am not using a venv at all.

The value that was there was `const PYTHONHOME = "/home/jriesen/.pyenv/versions/3.10.8:/home/jriesen/.pyenv/versions/3.10.8"` which is basically correct, that's where my currently-active Python install is located, so it did pick up the path correctly, it just seems that actually specifying it breaks everything.
